syncer/syncer.go

```diff
@@ -654,17 +671,26 @@ func (s *Syncer) syncLoop(ctx context.Context) error {
 // connections, and syncing the blockchain from active peers. It blocks until an
 // error occurs, upon which all connections are closed and goroutines are
 // terminated.
-func (s *Syncer) Run(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	s.wg.Add(1)
-	defer s.wg.Done()
+func (s *Syncer) Run() error {
+	ctx, cancel, err := s.tg.AddContext(context.Background())
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
 	errChan := make(chan error)
-	s.wg.Add(3)
-	go func() { errChan <- s.acceptLoop(ctx); s.wg.Done() }()
-	go func() { errChan <- s.peerLoop(ctx); s.wg.Done() }()
-	go func() { errChan <- s.syncLoop(ctx); s.wg.Done() }()
-	err := <-errChan
-	cancel()
+	for _, fn := range []func(context.Context) error{s.acceptLoop, s.peerLoop, s.syncLoop} {
+		go func() {
+			done, err := s.tg.Add()
+			if err != nil {
+				errChan <- err
+				return
+			}
+			errChan <- fn(ctx)
```

Comment:
Fine to leave as is but I'd make `errChan` buffered. I had to manually verify whether you follow this up with two `<-errChan`, otherwise we'd potentially be deadlocked again.

Reply:
Making it buffered breaks the syncer cause it would not block on the first receive leading to the listener being closed immediately.